### PR TITLE
ordered smoothing of categorical features

### PR DESCRIPTION
**Pull request by FelixWick**

### CODE REVIEW

File: cyclic_boosting/utils.py [hunk 1]
```diff
@@ -834,14 +834,54 @@ def regularize_to_error_weighted_mean(values, uncertainties, prior_prediction=No
...
     else:
         if np.allclose(x, prior_prediction):
             return x
         x_mean = prior_prediction
     wx_incl = 1.0 / (np.sum(wx * np.square(x - x_mean)) / sum_wx)
     res = (wx * x + wx_incl * x_mean) / (wx + wx_incl)
+    return res
+
+
+def regularize_to_error_weighted_mean_neighbors(values, uncertainties, window_size=3):
```
Comment (lbventura) [inline cyclic_boosting/utils.py:847]: Add return type `np.ndarray`

File: cyclic_boosting/utils.py [hunk 2]
```diff
@@ -834,14 +834,54 @@ def regularize_to_error_weighted_mean(values, uncertainties, prior_prediction=No
...
+    """
+    if values.shape != uncertainties.shape:
+        raise ValueError("values and uncertainties must have the same shape")
+
+    if len(values) < 3:
+        return regularize_to_error_weighted_mean(values, uncertainties)
+
+    window_arr = np.ones(window_size)
+    x = values
+    wx = 1.0 / np.square(uncertainties)
```
Comment (lbventura) [inline cyclic_boosting/utils.py:875]: Is there a check upstream that the uncertainties are finite or could one of the values in wx in principle be `inf`?

File: cyclic_boosting/utils.py [hunk 3]
```diff
@@ -834,14 +834,54 @@ def regularize_to_error_weighted_mean(values, uncertainties, prior_prediction=No
...
+
+    sum_wx = np.convolve(wx, window_arr, "same")
+    x_mean = np.convolve(wx * x, window_arr, "same") / sum_wx
+
+    wx_incl = np.ones(len(x))
+    for i in range(len(x)):
+        wx_incl[i] = 1.0 / (np.convolve(wx * np.square(x - x_mean[i]), window_arr, "same") / sum_wx)[i]
```
Comment (lbventura) [inline cyclic_boosting/utils.py:882]: Pedantic (and possible wrong) remark, but is this not the same as 
```suggestion
        wx_incl[i] = (sum_wx/np.convolve(wx * np.square(x - x_mean[i]), window_arr, "same"))[i]
```

File: tests/test_utils.py [hunk 4]
```diff
@@ -73,3 +73,23 @@ def test_get_normalized_values(values, expected_result):
         np.testing.assert_almost_equal(sum(normalized_values), 1.0, 6)
     else:
         np.testing.assert_almost_equal(sum(normalized_values), 0.0, 6)
+
+
+def test_regularize_to_error_weighted_mean_neighbors():
+    X = pd.DataFrame({"a": [0, 1], "b": [3, 6], "c": [2, 7]})
```
Comment (lbventura) [inline tests/test_utils.py:79]: Why are these lines part of the test?
```
    X = pd.DataFrame({"a": [0, 1], "b": [3, 6], "c": [2, 7]})
    features = utils.get_feature_column_names(X)
    np.testing.assert_equal(features, ["a", "b", "c"])
    features = utils.get_feature_column_names(X, exclude_columns=["a"])
    np.testing.assert_equal(features, ["b", "c"])
```
It seems like there are two functions `get_feature_column_names` and `regularize_to_error_weighted_mean_neighbors` being tested here.

File: tests/conftest.py [hunk 5]
```diff
@@ -38,10 +38,10 @@ def feature_properties() -> dict:
     fp["P_ID"] = flags.IS_UNORDERED
     fp["PG_ID_3"] = flags.IS_UNORDERED
     fp["L_ID"] = flags.IS_UNORDERED
-    fp["dayofweek"] = flags.IS_ORDERED
+    fp["dayofweek"] = flags.IS_UNORDERED
```
Comment (lbventura) [inline tests/conftest.py:41]: Ignorant question, why are "dayofweek" and "PROMOTION_TYPE" unordered? I would have thought that the former is ordered

**Review by lbventura — APPROVED** [review]

File: tests/conftest.py (hunk 5 above)
Comment (FelixWick) [reply tests/conftest.py:41]: Yes, you're right, day of week is ordered. However, it is a cyclical ordering, i.e., day 0 should be smoothed by day 6. And this is not done (intentionally) with the implementation of ordered smoothing here. So, I think unordered is better.

File: cyclic_boosting/utils.py (hunk 2 above)
Comment (FelixWick) [reply cyclic_boosting/utils.py:875]: Yeah, I should check for 0 👍.

**Pull request merged**